postgres/src/config.rs

```diff
@@ -308,32 +309,51 @@ impl Config {
         self.config.get_channel_binding()
     }
 
-    /// Opens a connection to a PostgreSQL database.
-    pub fn connect<T>(&self, tls: T) -> Result<Client, Error>
+    /// Like `connect`, but `notify_fn` will be invoked with any asynchronous
+    /// notifications or connection errors that arise.
+    ///
+    /// `notify_fn` is invoked from an asynchronous runtime and therefore must
+    /// not block.
+    pub fn connect_with<T, F>(&self, tls: T, mut notify_fn: F) -> Result<Client, Error>
```

Comment:
Rather than a separate method, I think it would make more sense to store the callback in the config with a setter.

Reply:
How do you propose to store the closure in the config? I see two options:

  * An `fn (AsyncFunction)` function pointer. This is almost certainly too limiting since you can't pass a closure that captures any state. (It's definitely too limiting for my uses.)
  * A `Box<dyn FnMut(AsyncMessage) + Clone + Send + 'static>`. This doesn't work as written, of course, because `Clone` isn't an auto trait, but I think you could make it work by creating a child trait that inherits from `Clone` and `FnMut(AsyncMessage)`? (`Clone` is necessary to preserve the `Clone`ness of `Config`). This is also potentially limiting, if the closure wants to capture a non-`Clone`able type.

Reply:
I'd go with an `Arc<dyn Fn(AsyncMessage) + Sync + Send>`.

Reply:
Forcing a `Sync` bound on the closure is tough, though. For example it precludes the test that I wrote which sends the messages over an MPSC (which I imagine is what 90% of these implementations will do) because an MPSC transmitter is not `Sync`.  And putting an MPSC transmitter in a `Mutex` feels really weird, especially because we know that each closure is only going to be invoked by one thread.

I've pushed up an implementation that instead uses a `Box<NotificationCallback>`, where `NotificationCallback` is a custom private trait that's implemented for any `FnMut(AsyncMessage) + Send + Clone`. From the user's perspective the API is

```rust
    pub fn notification_callback<F>(&mut self, f: F) -> &mut Config
    where
        F: Fn(AsyncMessage) + Clone + Send + 'static,
    {
```

which feels quite nice. A bit of ugly boilerplate internally though.